sys/kern/kern_exec.c

```diff
@@ -1173,16 +1190,31 @@ exec_new_vmspace(struct image_params *imgp, struct sysentvec *sv)
 #endif
 	p->p_psstrings = p->p_usrstack - sv->sv_szpsstrings;
 
+	/* We reserve the whole max stack size with restricted permission */
 	stack_addr =  p->p_usrstack - ssiz;
+	stack_prot = (obj != NULL && imgp->stack_prot != 0) ? imgp->stack_prot :
+	    sv->sv_stackprot;
 	imgp->stack_sz = ssiz;
-	error = vm_map_stack(map, stack_addr, (vm_size_t)ssiz,
-	    obj != NULL && imgp->stack_prot != 0 ? imgp->stack_prot :
-	    sv->sv_stackprot, VM_PROT_ALL, MAP_STACK_GROWS_DOWN);
+	error = vm_map_reservation_create(map, &stack_addr, ssiz,
+	    PAGE_SIZE, stack_prot);
 	if (error != KERN_SUCCESS)
 		return (vm_mmap_to_errno(error));
+
+	error = vm_map_stack(map, stack_addr, (vm_size_t)ssiz, stack_prot,
+	    stack_prot, MAP_STACK_GROWS_DOWN);
+	if (error != KERN_SUCCESS) {
+		vm_map_reservation_delete(map, stack_addr);
+		return (vm_mmap_to_errno(error));
+	}
+
 #if __has_feature(capabilities)
-        imgp->stack = cheri_capability_build_user_data(
-	    CHERI_CAP_USER_DATA_PERMS, stack_addr, ssiz, ssiz);
+	perms = CHERI_CAP_USER_DATA_PERMS & vm_map_prot2perms(stack_prot);
```

Comment:
This seems to be stripping the global perm?

Reply:
It looks like so, need to fix this.